bench/wtperf/wtperf.c

```diff
@@ -131,6 +146,177 @@ randomize_value(CONFIG_THREAD *thread, char *value_buf)
 }
 
 static int
+setup_truncate(CONFIG *cfg,
+    CONFIG_THREAD *thread, TRUNCATE_CONFIG *trunc, WT_SESSION *session) {
+
+	TRUNCATE_QUEUE_ENTRY *truncate_item;
+	WT_CURSOR *cursor;
+	size_t i;
+	uint64_t end_point_val, final_milestone_gap, start_point_val;
+	char *key, *truncate_key;
+	int ret;
+
+	ret = 0;
+	end_point_val = final_milestone_gap = start_point_val = 0;
+
+	/* We are limited to only one table when running truncate */
+	ret = session->open_cursor(session, cfg->uris[0], NULL, NULL, &cursor);
+	if (ret != 0) {
+		return (ret);
+	}
+
+	/* Truncation percentage value. eg 10% is 0.1 */
+	trunc->truncation_percentage =
+	    (double)thread->workload->truncate_pct / 100;
+	/* How many entries between each milestone */
+	trunc->truncate_milestone_gap =
+	    thread->workload->truncate_count * trunc->truncation_percentage;
+	/* How many milestones we need */
+	trunc->needed_milestones = thread->workload->truncate_count /
+	    trunc->truncate_milestone_gap;
+
+	final_milestone_gap = trunc->truncate_milestone_gap;
+
+	/* Reset this value for use again */
+	trunc->truncate_milestone_gap = 0;
+
+	/*
+	 * Here we check if there is data in the collection. If there is
+	 * data available, then we need to setup some initial truncation
+	 * milestones.
+	 */
+	ret = cursor->next(cursor);
+
+	if (ret == 0) {
+		ret = cursor->get_key(cursor, &key);
+	}
+	/* We have data */
+	if (ret == 0) {
+		start_point_val = decode_key(cfg, key);
+		cursor->reset(cursor);
+		if ((ret = cursor->prev(cursor)) != 0) {
+			lprintf(cfg, ret, 0,
+			    "truncate setup end: prev failed");
+			return (ret);
```

Comment:
Each return after the cursor is opened should be replaced with a ```goto err;```, and the error cleanup should close the cursor.